sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rule/BeamIOPushDownRule.java

```diff
@@ -124,11 +126,20 @@ public void onMatch(RelOptRuleCall call) {
     RelDataType calcInputType =
         CalciteUtils.toCalciteRowType(newSchema, ioSourceRel.getCluster().getTypeFactory());
 
+    // TODO: Check if an IO supports field reordering and drop a Calc when it does (1).
     // Check if the calc can be dropped:
-    // 1. Calc only does projects and renames.
+    // 1. Calc only does projects and renames of fields in the same order.
     //    And
     // 2. Predicate can be completely pushed-down to IO level.
-    if (isProjectRenameOnlyProgram(program) && tableFilter.getNotSupported().isEmpty()) {
+    //    And
+    // 3. And IO supports project push-down OR all fields are projected by a Calc.
+    if (isProjectRenameOnlyProgram(program, beamSqlTable.supportsProjects())
```

Comment:
Another case when `Calc` should not be dropped is when an IO does not support field reordering and fields are not projected in the same order they are listed in the IO Schema.
Did not get cough by tests, because TestTable providers uses Select, which does field reordering for us.

Reply:
This should be resolved now, IOs can specify whether they support field reordering. When reordering in not supported and fields are projected in a different order (from Schema) - Calc should not be dropped.